### PR TITLE
ECC enhancements

### PR DESCRIPTION
**Pull request by karel-m**
This PR is related to https://github.com/libtom/libtomcrypt/issues/34

### CODE REVIEW

**Comment by sjaeckel** [comment]
I think the build error can be fixed in several ways:

1. by making point compression optional and disabling it by default
2. creating a new release of libtommath with the required additional algorithm, waiting until it's published
3. building libtommath from the git repositories
4. combining 1&3 and creating a special recipe with point compression enabled and build ltm from git

I would prefer going for 2 or 4

**Comment by karel-m** [comment]
Or we can introduce in libtommath something like LTM_VERSION (e.g. 0x0042 for version 0.42) and then in libtomcrypt's `ltm_desc.c`:

```c
/* sqrtmod_prime */
static int sqrtmod_prime(void *a, void *b, void *c)
{
   LTC_ARGCHK(a != NULL);
   LTC_ARGCHK(b != NULL);
   LTC_ARGCHK(c != NULL);
#if !defined(LTM_VERSION) || LTM_VERSION < 0x0042
   return CRYPT_LTM_TOO_OLD; 
#else
   return mpi_to_ltc_error(mp_sqrtmod_prime(a, b, c));
#endif
}
```

**Comment by karel-m** [comment]
@fperrad may I ask you for running your linting machine on `miko-ecc-enhancements` branch? Thanks.

**Comment by fperrad** [comment]
@karel-m linting issues:

missing prototypes in tomcrypt.h : ecc_sign_hash_ex & ecc_verify_hash_ex
```
Info 765: external 'ecc_sign_hash_ex(const unsigned char *, unsigned long, unsigned char *, unsigned long *, union Prng_state *, int, struct ecc_key *, int)' (line 24, file src/pk/ecc/ecc_sign_hash.c) could be made static
Info 765: external 'ecc_verify_hash_ex(const unsigned char *, unsigned long, const unsigned char *, unsigned long, int *, struct ecc_key *, int)' (line 34, file src/pk/ecc/ecc_verify_hash.c) could be made static
```
bad indentation line 30 in src/pk/ecc/ecc_export_raw.c

bad indentation line 62 in src/pk/ecc/ecc_sign_hash.c

bad indentation line 72, 74 & 75 in src/pk/ecc/ecc_import_raw.c

**Comment by sjaeckel** [comment]
can we somehow bring the coverage up again?

**Comment by karel-m** [comment]
> can we somehow bring the coverage up again?

What has to be done to bring it up?

**Comment by sjaeckel** [comment]
1. These aren't tested at all

        ecc_sign_hash_rfc7518()
        ecc_verify_hash_rfc7518()

1. There are no tests for all the added functionality

        pk_get_oid()

1. The new `tag` field in DER/`ltc_asn1_list`

1. The case `a==NULL` in some of the code paths that are handled specially

**Comment by karel-m** [comment]
Oh yes, tests, I know.

File: src/pk/asn1/der/sequence/der_encode_sequence_ex.c [hunk 1]
```diff
@@ -207,141 +88,145 @@ int der_encode_sequence_ex(ltc_asn1_list *list, unsigned long inlen,
...
-               x       += z;
-               *outlen -= z;
                break;
 
            case LTC_ASN1_CHOICE:
            case LTC_ASN1_CONSTRUCTED:
            case LTC_ASN1_CONTEXT_SPECIFIC:
            case LTC_ASN1_EOL:
            case LTC_ASN1_TELETEX_STRING:
+           default:
```
Comment (sjaeckel) [inline src/pk/asn1/der/sequence/der_encode_sequence_ex.c]: I never use a default case when I switch on an enum as the compiler then warns me if I add elements to the enum

Comment (sjaeckel) [inline src/pk/asn1/der/sequence/der_encode_sequence_ex.c]: I prefer not to use a default case when I switch on an enum as the compiler then warns me if I add elements to the enum and forget to modify the switch-case

**Comment by karel-m** [comment]
@sjaeckel this PR is too complex (rebasing is quite often a pain), I am considering the idea of splitting this PR to two parts: 1/ ASN.1 changes + 2/ EC-only stuf. 

The part 1/ would be a new PR (which I want to merge to develop before continuing on the remaining EC-only part) comprising of approx. ede958b + subset of 01c319d 

What do you think?

**Comment by sjaeckel** [comment]
> ...splitting this PR into two parts: 1/ ASN.1 changes + 2/ EC-only stuf.

Very good idea, that also makes review easier. Can the EC part probably be split even more?

**Comment by karel-m** [comment]
> Can the EC part probably be split even more?

We'll see, let's start with ASN.1 part

**Comment by sjaeckel** [comment]
We need a ltm-1.0 package backported to trusty to fix these build failures.
I had to remove the ltm from sid in #217 as I couldn't update clang otherwise...

**Comment by karel-m** [comment]
I have split this PR and my branch `miko-ecc-enhancements` to:
* #236 ECC enhancements part 1 (curves y^2 = x^3 + ax + b) - branch `pr/ecc-non-asn1-part`
* #187 ECC enhancements part 2 (ASN.1 changes) - branch `pr/ecc-asn1-part` (based on `pr/ecc-non-asn1-part`)

So basically current `miko-ecc-enhancement` is the same as `pr/ecc-asn1-part`

There fore I am closing this PR, please move futher discussion to #236 + #187

**Pull request closed without merging**